runtime/src/cost_tracker.rs

```diff
@@ -199,40 +222,64 @@ impl CostTracker {
 
     fn add_transaction_cost(&mut self, tx_cost: &TransactionCost) {
         let cost = tx_cost.sum();
+        self.add_transaction_execution_cost(tx_cost, cost);
+        saturating_add_assign!(self.account_data_size, tx_cost.account_data_size);
+        saturating_add_assign!(self.transaction_count, 1);
+    }
+
+    fn remove_transaction_cost(&mut self, tx_cost: &TransactionCost) {
+        let cost = tx_cost.sum();
+        self.sub_transaction_execution_cost(tx_cost, cost);
+        self.account_data_size = self
+            .account_data_size
+            .saturating_sub(tx_cost.account_data_size);
+        self.transaction_count = self.transaction_count.saturating_sub(1);
+    }
+
+    /// Apply additional actual execution units to cost_tracker
+    fn add_transaction_execution_cost(&mut self, tx_cost: &TransactionCost, adjustment: u64) {
         for account_key in tx_cost.writable_accounts.iter() {
             let account_cost = self
                 .cost_by_writable_accounts
                 .entry(*account_key)
                 .or_insert(0);
-            *account_cost = account_cost.saturating_add(cost);
+            *account_cost = account_cost
+                .checked_add(adjustment)
+                .unwrap_or(self.account_cost_limit);
         }
-        self.block_cost = self.block_cost.saturating_add(cost);
+        self.block_cost = self
+            .block_cost
+            .checked_add(adjustment)
+            .unwrap_or(self.block_cost_limit);
```

Comment:
I don't like that overflow and underflow sets the tracked cost to the limit but subsequent cost adjustments can move the cost away from the limit again. I think that once an unexpected underflow/overflow occurs, it should be impossible for the costs to be adjusted again.

Reply:
> I don't like that overflow and underflow sets the tracked cost to the limit but subsequent cost adjustments can move the cost away from the limit again. I think that once an unexpected underflow/overflow occurs, it should be impossible for the costs to be adjusted again.

Hmm, that should not happen. Once tracked numbers (account, block, vote etc) set to their limits, `cost_tracker.try_add(...)` will return error, no transaction will be selected for block inclusion, hence no more adjustment to tracker.

Reply:
What if you remove 2 transactions from the cost tracker and the first to removal causes underflow. The second removal could bring all the costs below the limit again, right?

Reply:
right!

Reply:
might be time for an enum.  we're encoding too much

Reply:
> What if you remove 2 transactions from the cost tracker and the first to removal causes underflow. The second removal could bring all the costs below the limit again, right?

Can't find a cleaner way to handle over/under flow with `checked_` math. Need back to drawing board to define what to do with unexpected adjustment. For now, go back to use `saturating_` in [commit](a1b30b3034a46f91de4aa9831e33c110f0545bb8).

Reply:
> might be time for an enum. we're encoding too much

Good call. Prolly can define something like this: 
```
enum TrackedValued {
    ComputeUnit(u64),
    DataSize(u64),
    TransactionCount(u64),
    /// Value is immutable due to unexpected error, eg over/under flow
    Frozen, 
    /// Value had reached its max limit
    AtLimit,
}
```
Then we can handle add/sub corner cases, would you think be OK to create an issue to refactor it separately, or necessary to do it in this PR?

Reply:
I'm happy with saturating arithmetic for now

Reply:
Kool, thanks! Issue to refactor with enum: https://github.com/solana-labs/solana/issues/24551